solvers/H3LAPD/EquationSystems/DriftReducedSystem.cpp

```diff
@@ -0,0 +1,578 @@
+#include <LibUtilities/TimeIntegration/TimeIntegrationScheme.h>
+#include <boost/core/ignore_unused.hpp>
+
+#include "DriftReducedSystem.hpp"
+
+namespace NESO::Solvers::H3LAPD {
+DriftReducedSystem::DriftReducedSystem(
+    const LU::SessionReaderSharedPtr &session,
+    const SD::MeshGraphSharedPtr &graph)
+    : UnsteadySystem(session, graph), AdvectionSystem(session, graph),
+      m_field_to_index(session->GetVariables()),
+      m_adv_vel_elec(graph->GetSpaceDimension()),
+      m_ExB_vel(graph->GetSpaceDimension()), m_E(graph->GetSpaceDimension()) {
+  // Construct particle system
+  m_particle_sys = std::make_shared<NeutralParticleSystem>(session, graph);
+}
+
+/**
+ * @brief Compute advection terms and add them to an output array
+ * @details For each field listed in @p field_names copy field pointers from
+ * m_fields and physical values from @p in_arr to temporary arrays and create a
+ * temporary output array of the same size. Call Advect() on @p adv_obj passing
+ * the temporary arrays. Finally, loop over @p eqn_labels to determine which
+ * element(s) of @p out_arr to subtract the results from.
+ *
+ * N.B. The creation of temporary arrays is necessary to bypass restrictions in
+ * the Nektar advection API.
+ *
+ * @param field_names List of field names to compute advection terms for
+ * @param adv_obj Nektar advection object
+ * @param adv_vel Array of advection velocities (outer dim size = nfields)
+ * @param in_arr Physical values for *all* fields
+ * @param[out] out_arr RHS array (for *all* fields)
+ * @param time Simulation time
+ * @param eqn_labels List of field names identifying indices in out_arr to add
+ * the result to. Defaults to @p field_names
+ */
+void DriftReducedSystem::add_adv_terms(
+    std::vector<std::string> field_names, const SU::AdvectionSharedPtr adv_obj,
+    const Array<OneD, Array<OneD, NekDouble>> &adv_vel,
+    const Array<OneD, const Array<OneD, NekDouble>> &in_arr,
+    Array<OneD, Array<OneD, NekDouble>> &out_arr, const NekDouble time,
+    std::vector<std::string> eqn_labels) {
+
+  // Default is to add result of advecting field f to the RHS of df/dt equation
+  if (eqn_labels.empty()) {
+    eqn_labels = std::vector(field_names);
+  } else {
+    ASSERTL1(field_names.size() == eqn_labels.size(),
+             "add_adv_terms: Number of quantities being advected must match "
+             "the number of equation labels.");
+  }
+
+  int nfields = field_names.size();
+  int npts = in_arr[0].size();
+
+  /* Make temporary copies of target fields, in_arr vals and initialise a
+   * temporary output array
+   */
+  Array<OneD, MR::ExpListSharedPtr> tmp_fields(nfields);
+  Array<OneD, Array<OneD, NekDouble>> tmp_inarray(nfields);
+  Array<OneD, Array<OneD, NekDouble>> tmp_outarray(nfields);
+  for (auto ii = 0; ii < nfields; ii++) {
+    int idx = m_field_to_index.get_idx(field_names[ii]);
+    tmp_fields[ii] = m_fields[idx];
+    tmp_inarray[ii] = Array<OneD, NekDouble>(npts);
```

Comment:
Does the solver assert somewhere that all the fields have the same order - hence same number of quadrature points (if this assumption is throughout?)?

Reply:
It doesn't, good point.
I've added a check in DriftReducedSystem::validate_fields().